Yoast/wordpress-seo
# Add the last twitter title presenters

## Summary


This PR can be summarized in the following changelog entry:

* N/A

## Relevant technical choices:

* Implemented `twitter_title`, including unit tests, for the following page types:
    * Attachment
    * Author archive
    * Date archive
    * Search results
* Decided to change the fallback logic of the author and date archives. Previously, this used complex logic from `class-frontend` (see [here](https://github.com/Yoast/wordpress-seo/blob/7205b0fa8f0141221d014b708e45a7c2012ca5e4/frontend/class-frontend.php#L343) for author archives, and see [here](https://github.com/Yoast/wordpress-seo/blob/7205b0fa8f0141221d014b708e45a7c2012ca5e4/frontend/class-frontend.php#L515) for date archives). This code was old and seems to have been replaced by the title options default in `class-wpseo-option-titles`. Therefore, for empty titles we now default to `class-wpseo-option-titles`.

## Test instructions

This PR can be tested by following these steps:

* The page source should show an indented `twitter-title` tag with the correct title. This is the example for a simple post:
<img width="695" alt="Screenshot 2019-10-01 at 15 02 07" src="https://user-images.githubusercontent.com/20280513/65964144-8fab4080-e45c-11e9-922a-452882d026f4.png">

* When testing, (also) use snippet variables in your `twitter-title`, for example `%%sitename%%`. Make sure they are output correctly.

* You should check whether the `twitter-title` tag is present for the following types of content
    * Attachment page
        * In the menu on the left, click `SEO` --> `Search appearance` --> `Media`
        * Set `Redirect attachment URLs to the attachment itself?` to `No`, set `Yoast SEO Meta Box` to `Show`, then click `Save changes`
        * In the menu on the left, click `Media`
        * Click on an image (or first add an image if the media library is empty), then click `View attachment page`
        * In the top menu, click `Edit media`. Give the attachment page an SEO title (metabox: `Edit snippet` --> `SEO title`), and a Twitter title (metabox --> `Social` --> add a `Twitter title`). Then click `Update`
        * View the attachment page again (it seems that for this, you need to follow the first two above steps again)
        * The `twitter:title` content is expected to be the Twitter title you entered in the metabox.
        * Remove the `Twitter title` from the metabox and view the attachment page again.
        * The `twitter:title` content is expected to be the SEO title.
        * Remove the `Twitter title` from the metabox and view the attachment page again.
        * The `twitter:title` content is expected to be the Yoast SEO installation default: `%%title%% %%page%% %%sep%% %%sitename%%` (`%%title%%` is the post's 'normal' title).
    * Author archives
        * In the menu on the left, click `Users` --> `Your profile` --> `Title to use for Author page`
        * Fill in some nice title, then click `Update profile`
        * In the menu on the left, click `SEO` --> `Search appearance` --> `Archives` --> `Author archive settings`
        * Change the `SEO title` to something (maybe add some replacement variables), then click `Save changes`
        * Navigate to an author archive, for example `http://one.wordpress.test/author/admin/` (potentially replace `one` with the name of your installation)
        * The `twitter:title` content is expected to be the title you entered on the user profile.
        * In the menu on the left, click `Users` --> `Your profile` --> `Title to use for Author page`
        * Remove the text from the title, then click `Update profile`
        * The `twitter:title` content is expected to be the title you entered on the search appearance settings.
        * In the menu on the left, click `SEO` --> `Search appearance` --> `Archives` --> `Author archive settings`
        * Remove the text from the `SEO title`, then click `Save changes`
        * The `twitter:title` content is expected to be the Yoast SEO installation default: `%%name%%, Author at %%sitename%% %%page%%` (`Author at` is translated in other languages).
    * Date archives
        * In the menu on the left, click `SEO` --> `Search appearance` --> `Archives` --> `Date archive settings`
        * Change the `SEO title` to something (maybe add some replacement variables), then click `Save changes`
        * Navigate to a date archive, for example `http://one.wordpress.test/2019/09/13/`
        * The `twitter:title` content is expected to be the title you entered on the search appearance settings. 
        * Check again with an empty `SEO title`
        * The `twitter:title` content is expected to be the Yoast SEO installation default: `%%date%% %%page%% %%sep%% %%sitename%%`.
    * Search result page
        * In the menu on the left, click `SEO` --> `Search appearance` --> `Archives` --> `Special pages` --> `Search pages`
        * Change the `SEO title` to something (maybe add some replacement variables), then click `Save changes`
        * Navigate to a search result page, for example `http://one.wordpress.test/?s=gutenberg`
        * The `twitter:title` content is expected to be the title you entered on the search appearance settings.
        * Check again with an empty `SEO title`
        * The `twitter:title` content is expected to be the Yoast SEO installation default: `You searched for %%searchphrase%% %%page%% %%sep%% %%sitename%%`.

## UI changes

## Documentation

## Quality assurance


Fixes #
